### DISCUSSION TITLE
[XState v5 beta] Restoring state of spawned actors from persisted state not working

### DISCUSSION THREAD

**Discussion by Wastl77** [statelyai/xstate#4171] [Q&A] [answered]
Hi,

i have a problem with restoring state of spawned actors. I have a root machine, which invokes a fetchInitialData machine.  When the data is fetched, it sends an event to the root machine with the fetched data:

```
on: {
  'FETCH-SUCCESS': {
    actions: [
	assign({ fetchedResources: ({ event }) => event.data }),
    ],
    target: 'spawnResources',
  },
```

In the spawnResources-State the data from context will spawn the resources:

```
spawnResources: {
  entry: [
     assign({
      ref: ({ context, spawn }) =>
        context.fetchedResources.map((res: any) => {
          return spawn(resource, {
            systemId: res.callsign,
          });
        }),
    }),
  ],
  always: { target: 'ready' },
},
```

When i now persist the state from the root machine, the persisted state looks like this (shortened for readability, only on resource):

`{"value":"ready","done":false,"context":{"fetchedResources":[{"id":"64b7aeb3f69aa8b39da5c5c8","callsign":"1182-1"}],"ref":[{"id":"x:4"}],"isSession":true},"historyValue":{},"_internalQueue":[],"children":{"x:4":{"state":{"value":{"available State":"D1"},"done":false,"context":{},"historyValue":{},"_internalQueue":[],"children":{}}}}}`

Note the children property in the JSON, it has a state with value "available State: D1". When restoring the state with:

```
const persistedState = await getPersistedState();

    rootActor = interpret(notbetriebRootMachine, {
      systemId: 'Notbetrieb Root',
      state: persistedState,
    });
rootActor.start();
```

If persisting the state again from the newly started machine interpreter, the JSON looks as following:

`{"value":"ready","done":false,"context":{"fetchedResources":[{"id":"64b7aeb3f69aa8b39da5c5c8","callsign":"1182-1"}],"ref":[{"id":"x:4"}],"isSession":true},"historyValue":{},"_internalQueue":[],"children":{}}`

Here you can see, the ref to the spawned actor is there, but in the children property, no state is stored. It also doesn't work to set a new state in the actor in the running interpreter. So i'm assuming the spawned actor state is not recursively restored.

Note that i'm not invoking the fetchInitialData machine when restored state is present, but going directly to the spawnResources State:

```
idle: {
        always: [
          {
            guard: ({ context }) => context.isSession === true,
            target: 'spawnResources',
          },
          { target: 'fetchInitialData' },
        ],
      },
```

Any ideas why this is not working? Thank you!

**Comment by davidkpiano** [comment, marked as answer]
This is likely going to be solved here: #3921

**Reply by Wastl77** [reply]
OK, this would be great! Thank you for this great library!
